[Feature] support sequence parallelism using compilation pass

cascade812 opened the pull request:
This PR support sequence parallelism using below compilation config
```
config = CompilationConfig(
    level=3,
    custom_ops=["+rms_norm"],
    compile_sizes=[4, 8, 16],
    splitting_ops=[],
)
config.pass_config.enable_sequence_parallelism= True

llm = LLM(model="llama/Llama-3.2-1B-Instruct",
          enforce_eager=False,
          tensor_parallel_size=2,
          dtype=torch.float16,
          max_num_batched_tokens=2048,
          compilation_config=config)
```

- [x] Integrate comm/comp overlapping library https://github.com/vllm-project/vllm/pull/17882
- [x] Support pipeline parallelism https://github.com/vllm-project/vllm/pull/18243
- [ ] Support quantization 
- [ ] Test more models

cascade812 committed:
add reduce scatter op and register all gather

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
replace all reduce with reduce scatter and all gather

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
match first embedding

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
update embedding replace pattern

Signed-off-by: cascade812 <cascade812@outlook.com>

robertgshaw2-redhat commented:
nice!

cascade812 committed:
compile graph only for specific shapes

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
clean code

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
add test and rename

Signed-off-by: cascade812 <cascade812@outlook.com>

yaochengji commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,307 @@
+        )
+
+        gemm_1_weights = torch.empty([4, 4],
+                                     device="cuda",
```
Thanks for your great contribution! Currently I only have one comment. And I'm waiting for @tlrmchlsmth 's comprehensive review!

I think this pass can be hardware agnostic, so the device should not be hardcoded as "cuda".

cascade812 replied:
thanks @yaochengji !  there're other passes in `fusion.py` which also use 'cuda' directly.  I tried to read device by `get_current_vllm_config().device_config.device`, however `device_config` is none when `register_replacement` happens. I'm wondering if you know how to get device that would be suitable for this situation, thanks!

yaochengji replied:
From the pytorch code, https://github.com/pytorch/pytorch/blob/b0bdd76f2e410d8411febc6cab9d2d3bea9fc4b4/torch/_inductor/pattern_matcher.py#L1744, it looks like the device type should be ignored in the pattern match. Does it work when you use cpu device?

cascade812 replied:
I tried to remove device type or specify device to 'cpu', both doesn't work out. 
this is error when specified 'cpu'. 
`NotImplementedError: could not find kernel for vllm.all_reduce.default at dispatch key DispatchKey.CPU`

youkaichao commented on vllm/config.py:
```diff
@@ -3706,6 +3718,14 @@ def _set_cudagraph_sizes(self):
+                # enable sequence parallelism
+                if self.parallel_config.tensor_parallel_size > 1 and \
+                    self.compilation_config.pass_config.enable_sequence_parallelism:
+                    batch_size_capture_list = [
```
@bnellnm told me that flux kernel actually has more complicated requirements, something like `batchsize % (tp_size * 128) == 0` , not sure how different communication backends have this kind of alignment limitations.

cascade812 replied:
thanks for this info. this PR haven't integrated flux kernel yet. Plan to do it in next PR where we can discuss more about this.

tlrmchlsmth replied:
SGTM, we can use SP for other cases than just the flux kernels as well

youkaichao commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,307 @@
+    return rmsnorm[1], rmsnorm[2]
+
+
+def replace_with_gemm_rs_ag_rmsnorm(
```
does it need to be coupled with rms norm? is it possible to only match embedding + allreduce?

cascade812 replied:
yes, it's needed. because reduce scatter has to be put after rms norm.

youkaichao commented on vllm/compilation/compiler_interface.py:
```diff
@@ -291,11 +293,12 @@ def _get_shape_env() -> AlwaysHitShapeEnv:
-                example_inputs,
-                inner_compile=hijacked_compile_fx_inner,
-                config_patches=current_config)
+            with pass_context(runtime_shape):
```
here you have the runtime shape information, and you can create a copy of the `current_config`, remove the sequence parallel pass if size is not as expected. this way, `compile_fx` will get different `current_config` for different sizes.

cascade812 replied:
thanks for this idea. In current PR, I check if size is expected or not within the `extra_check` function when calling `register_replacement` in the `SequenceParallelismPass` class. I believe this approach keeps the logic more self-contained within the `SequenceParallelismPass`, which makes the code organization more coherent though it has to do the pattern match before the extra size check. Could you let me know which way you think is better?

youkaichao replied:
I think we should make a new API for the pass interface, let it check if it can be specialized for a shape, and then the pass manager queries all the compilation passes, to consider only relevant passes.

doing it in the pattern match looks weird.

also cc @ProExpertProg for the pass manager related thing.

cascade812 replied:
@youkaichao @ProExpertProg addressed this comment, thanks!

ProExpertProg commented:
I will take a closer look on Tuesday but my initial thoughts are that I agree we should improve the pass interface. At the same time, I like that currently all passes conform to the PyTorch `CustomGraphPass` interface (`__call__(graph: fx.Graph)`) and so I would prefer to pass the additional information through a context like in this PR. Also, just checking if a shape is "supported" is not enough; sometimes the pass needs to know the actual runtime shape

I also think we should better separate the `vllm.config.CompilationConfig` and the `torch._inductor.config`. In my attention fusion pass I will need access to the full `CompilationConfig` (to access attention layers) inside a pass (and not just `PassConfig`), which would recreate a cycle.

We should decide if we want to re-instantiate passes for different runtime shapes.

cascade812 committed:
address comments

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented on tests/distributed/test_comm_ops.py:
```diff
@@ -47,6 +48,34 @@ def all_reduce_test_worker(
+    # it is important to delete the CUDA_VISIBLE_DEVICES environment variable
+    # so that each worker can see all the GPUs
+    # they will be able to set the device to the correct GPU
+    monkeypatch.delenv("CUDA_VISIBLE_DEVICES", raising=False)
+    device = torch.device(f"cuda:{rank}")
+    torch.cuda.set_device(device)
+    init_test_distributed_environment(tp_size, pp_size, rank,
+                                      distributed_init_port)
```
this function is using Ray remote functions (@ray.remote) which run in separate processes. I think these processes don't have direct access to the pytest fixture mechanism that runs in the main test process.

yaochengji reviewed (approved):
LGTM, thanks for the contribution!

tlrmchlsmth commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1000,7 +1000,16 @@ def execute_model(
+            tp_size = self.vllm_config.parallel_config.tensor_parallel_size
+            if self.vllm_config.compilation_config.pass_config. \
+                enable_sequence_parallelism and tp_size > 1:
+                import math
+                num_input_tokens = math.ceil(
+                    num_scheduled_tokens / tp_size) * tp_size
```
nit: We have a `round_up` util for this (you can import it from `vllm/utils.py`)
```
num_input_tokens = round_up(num_scheduled_tokens, tp_size)
```

tlrmchlsmth commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,304 @@
+logger = init_logger(__name__)
+
+
+def get_world_name() -> str:
+    return torch.distributed.group.WORLD.group_name
+
+
+def residual_slice_shape(residual: torch.Tensor) -> int:
+    n_slices = get_tensor_model_parallel_world_size()
+    assert residual.size(0) % n_slices == 0
+    return residual.size(0) // n_slices
```
looks unused

tlrmchlsmth commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,304 @@
+        self.matches.append(match)
+        return bool(match)
+
+    def specialized_for_shape(self, shape: Optional[int]) -> bool:
+        # only do replace for specific shapes
+        tp_size = get_tensor_model_parallel_world_size()
+        return shape is not None and shape % tp_size == 0
```
unused?

tlrmchlsmth commented on vllm/distributed/parallel_state.py:
```diff
@@ -102,6 +102,14 @@ def _get_unique_name(name: str) -> str:
 _groups: Dict[str, Callable[[], Optional["GroupCoordinator"]]] = {}
 
 
+def get_group_from_group_name(group_name: str) -> "GroupCoordinator":
+    assert group_name in _groups, f"Group {group_name} is not found."
+    group = _groups[group_name]()
+    if group is None:
+        raise ValueError(f"Group {group_name} is destroyed.")
+    return group
```
unused?

tlrmchlsmth reviewed (commented):
Spotted a couple of unused functions that should be removed

tlrmchlsmth commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,304 @@
+    return normalized
+
+
+def generate_inputs_for_embedding_ar_rmsnorm():
+    arg2_1 = torch.rand([16, 4], device="cuda", dtype=torch.float16)
+    # mul_6: token indices (batch_size x seq_len)
+    mul_6 = torch.tensor([[3, 7, 1, 4, 9, 2, 5, 0]],
+                         device="cuda",
+                         dtype=torch.long)
+    unsqueeze = torch.rand([1, 8, 1], device="cuda", dtype=torch.float16) > 0.5
+    full_default = torch.zeros([1, 8, 4], device="cuda", dtype=torch.float16)
+    permute = torch.rand([1, 8, 4], device="cuda", dtype=torch.float16)
+    arg3_1 = torch.rand([4], device="cuda", dtype=torch.float16)
+    return [arg2_1, mul_6, unsqueeze, full_default, permute, arg3_1]
```
Do the values in these tensors matter at all?

cascade812 replied:
shapes and dtypes for those tensors do matter, values don't

tlrmchlsmth reviewed (commented):
PR looks very promising after a first pass, and a quite clean implementation. Thanks for the contribution! Left some minor comments. I'll need to look at the passes in more detail before accepting and I need to try it out as well!

tlrmchlsmth commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,287 @@
+    full_default = torch.zeros([1, 8, 4], device="cpu", dtype=torch.float16)
+    permute = torch.rand([1, 8, 4], device="cpu", dtype=torch.float16)
+    arg3_1 = torch.rand([4], device="cpu", dtype=torch.float16)
+    return [arg2_1, mul_6, unsqueeze, full_default, permute, arg3_1]
```
Looks like we could factor out the `torch.float16` and pass in a dtype so the pass could apply to both fp16 and bf16. (and then just run the pass twice).

Not sure if there's a better way to do it but I think that would be worth doing in this PR since bf16 is frequently the default dtype.

ProExpertProg replied:
FusionPass does this by parameteizing the patterns so they can be written once and reused

cascade812 replied:
parameterize dtype and device, and pass the value from vllm_config.

bnellnm commented on vllm/compilation/vllm_inductor_pass.py:
```diff
@@ -50,6 +51,9 @@ def end_and_log(self):
         duration_ms = float(self._end_time - self._start_time) / 1.0e6
         logger.debug("%s completed in %.1f ms", self.pass_name, duration_ms)
 
+    def is_applicable_for_shape(self, shape: Optional[int]):
+        return True
+
```
Why bother introducing this method unless `SequenceParallelismPass` has a specialized version of it?  Originally, this was used to check the flux shape requirements so if we aren't using flux yet it is not super important.

cascade812 replied:
this is needed for piecewise compilation where it doesn't support split the symbolic shape dimension. it apply the sequence parallelism only when shape is specific (for those shaped specified in compilation config)

tlrmchlsmth reviewed (commented):
I'm seeing some issues running the example in the summary:
```
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/vllm/compilation/backends.py", line 608, in __call__
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return self.compiled_graph_for_general_shape(*args)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 745, in _fn
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return fn(*args, **kwargs)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1184, in forward
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return compiled_fn(full_args)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 323, in runtime_wrapper
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     all_outs = call_func_at_runtime_with_args(
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/utils.py", line 126, in call_func_at_runtime_with_args
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     out = normalize_as_list(f(args))
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 672, in inner_fn
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     outs = compiled_fn(args)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 490, in wrapper
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return compiled_fn(runtime_args)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_inductor/output_code.py", line 466, in __call__
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return self.current_callable(inputs)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_inductor/utils.py", line 2128, in run
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return model(new_inputs)
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/.cache/vllm/torch_compile_cache/19057652a8/rank_0_0/inductor_cache/lh/clhw2i7qcliijf2cyx35q2takizctkzcvkkr5vlyfsqrc3slb753.py", line 250, in call
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     assert_size_stride(arg4_1, (s0, 2048), (2048, 1))
(VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383] AssertionError: expected size 1024==2048, stride 2048==2048 at dim=0
```

tlrmchlsmth commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,289 @@
+    embedding = torch.ops.aten.embedding.default(arg2_1, mul_6)
+    where = torch.ops.aten.where.self(unsqueeze, full_default, embedding)
+    all_reduce = tensor_model_parallel_all_reduce(where)
+    rmsnorm = torch.ops.higher_order.auto_functionalized(
+        torch.ops._C.rms_norm.default,
+        result=permute,
+        input=all_reduce,
+        weight=arg3_1,
+        epsilon=1e-5,
+    )
```
It seems that this only matches if we use the custom rmsnorm op.

From the summary in this PR, and from trying it myself, I see that `custom_ops=["+rms_norm"],` is necessary for this pass to do any fusion. Makes me think the passes are too brittle and we may want to return to that "noop sentinel" idea we were discussing, @yaochengji

yaochengji replied:
In my understanding, with `custom_ops=["+rms_norm"]`, it is wrapped as a higher_order op. And we only need to match the same higher_order op for different hardware implementations.

ProExpertProg replied:
This could be parametrized so that any rms_norm implementation (torch-based or custom) works here. Unless RMS doesn't need to be a part of the pattern?

bnellnm replied:
I tried using the torch based op here and it broke the pattern in other ways.  IIRC it was because pytorch did some CSE-like rewrites involving on one of the values used by the torch based rmsnorm.

ProExpertProg replied:
I see, then I guess sentinel it is?

bnellnm commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def generate_inputs_for_embedding_ar_rmsnorm():
+    arg2_1 = torch.rand([16, 4], device="cuda", dtype=torch.float16)
```
The types here probably matter since afaik, they are used for inference on the graph.  Maybe the types + device could be passed in as arguments?

The dimensions shouldn't really matter though.

cascade812 replied:
parameterize dtype and device, and pass the value from vllm_config.

bnellnm commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,289 @@
+            extra_check=lambda m: self.record_match(m),
+        )
+
+        gemm_1_weights = torch.empty([4, 4],
+                                     device="cuda",
+                                     dtype=torch.float16)
+        gemm_1_activations = torch.empty([4, 4],
+                                         device="cuda",
+                                         dtype=torch.float16)
+        residual = torch.empty([4, 4], device="cuda", dtype=torch.float16)
+        rms_norm_weights = torch.empty([4, 4],
+                                       device="cuda",
+                                       dtype=torch.float16)
```
The types here might need to be parameterized?

bnellnm commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,289 @@
+        embedding_match_cnt = self.embedding_ag_rmsnorm_pattern.apply(graph)
+        gemm_ar_rmsnorm_match_cnt = self.gemm_rs_ag_gemm_pattern.apply(graph)
+
+        if embedding_match_cnt > 0 or gemm_ar_rmsnorm_match_cnt > 0:
```
Should this condition be `and` instead of `or`?

bnellnm commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1000,7 +1000,15 @@ def execute_model(
-            num_input_tokens = num_scheduled_tokens
+            # Pad tokens to multiple of tensor_parallel_size when
+            # enabled collective fusion for SP
+            tp_size = self.vllm_config.parallel_config.tensor_parallel_size
+            if self.vllm_config.compilation_config.pass_config. \
+                enable_sequence_parallelism and tp_size > 1:
+                from vllm.utils import round_up
+                num_input_tokens = round_up(num_scheduled_tokens, tp_size)
```
This seems a bit fragile.  Especially if the constraints change, e.g. using flux kernels.

cascade812 replied:
we can make another function which takes other constraints into consideration when integrate with flux kernels.

cascade812 committed:
update

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented:
> I'm seeing some issues running the example in the summary:
> 
> ```
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/vllm/compilation/backends.py", line 608, in __call__
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return self.compiled_graph_for_general_shape(*args)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 745, in _fn
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return fn(*args, **kwargs)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1184, in forward
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return compiled_fn(full_args)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 323, in runtime_wrapper
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     all_outs = call_func_at_runtime_with_args(
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/utils.py", line 126, in call_func_at_runtime_with_args
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     out = normalize_as_list(f(args))
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 672, in inner_fn
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     outs = compiled_fn(args)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 490, in wrapper
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return compiled_fn(runtime_args)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_inductor/output_code.py", line 466, in __call__
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return self.current_callable(inputs)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/vllm/venv/lib/python3.10/site-packages/torch/_inductor/utils.py", line 2128, in run
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     return model(new_inputs)
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]   File "/home/tms/.cache/vllm/torch_compile_cache/19057652a8/rank_0_0/inductor_cache/lh/clhw2i7qcliijf2cyx35q2takizctkzcvkkr5vlyfsqrc3slb753.py", line 250, in call
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383]     assert_size_stride(arg4_1, (s0, 2048), (2048, 1))
> (VllmWorker rank=0 pid=1719091) ERROR 04-17 01:31:50 [multiproc_executor.py:383] AssertionError: expected size 1024==2048, stride 2048==2048 at dim=0
> ```

thanks for catching this. It was introduced by latest commits. Fixed.

yaochengji commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,304 @@
+    rms_norm_weights: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    gemm_1_w_perm = torch.ops.aten.permute.default(gemm_1_weights, [1, 0])
+    mm_1 = torch.ops.aten.mm.default(gemm_1_activations, gemm_1_w_perm)
```
Actually I don't think matmul is needed in the pattern match.

cascade812 replied:
removed permute and mm from pattern match.

cascade812 committed:
pass in dtype and device

Signed-off-by: cascade812 <cascade812@outlook.com>

tlrmchlsmth commented:
Dumb question, but how can I specify that compilation config when using `vllm serve` @youkaichao ?

tlrmchlsmth committed:
Merge branch 'main' into sp_pass

Signed-off-by: Tyler Michael Smith <tyler@neuralmagic.com>

bnellnm commented:
> Dumb question, but how can I specify that compilation config when using `vllm serve` @youkaichao ?

You should be able to use something like:
```
-O "{'level': 3, 'compile_sizes': [4, 8, 16], 'custom_ops': ['+rms_norm'], 'splitting_ops': []}"
```

tlrmchlsmth commented:
> > Dumb question, but how can I specify that compilation config when using `vllm serve` @youkaichao ?
> 
> You should be able to use something like:
> 
> ```
> -O "{'level': 3, 'compile_sizes': [4, 8, 16], 'custom_ops': ['+rms_norm'], 'splitting_ops': []}"
> ```

@bnellnm It needs the `config.pass_config.enable_sequence_parallelism = True` as well to turn on the sequence parallel pass

cascade812 committed:
enable rms_norm automatically if enable_sequence_parallelism=True

Signed-off-by: cascade812 <cascade812@outlook.com>

ProExpertProg commented:
> > > Dumb question, but how can I specify that compilation config when using `vllm serve` @youkaichao ?
> > 
> > You should be able to use something like:
> > 
> > ```
> > -O "{'level': 3, 'compile_sizes': [4, 8, 16], 'custom_ops': ['+rms_norm'], 'splitting_ops': []}"
> > ```
> 
> @bnellnm It needs the `config.pass_config.enable_sequence_parallelism = True` as well to turn on the sequence parallel pass



Tyler not sure if you're asking or adding to Bill's but here's the full one just in case it's the former:

```
-O "{'level': 3, 'compile_sizes': [4, 8, 16], 'custom_ops': ['+rms_norm'], 'splitting_ops': [], 'pass_config': {'enable_sequence_parallelism' : True}}"
```

cascade812 committed:
add test for sq pass

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
fix failed tests

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
fix failed tests

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 committed:
fix failed tests

Signed-off-by: cascade812 <cascade812@outlook.com>

tlrmchlsmth reviewed (approved):
LGTM when green, thanks for the excellent work

ProExpertProg commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,296 @@
+            arg3_1 = torch.rand([4], device=self.device, dtype=self.dtype)
+            return [arg2_1, mul_6, unsqueeze, full_default, permute, arg3_1]
+
+        pm.register_replacement(pattern, replacement, get_inputs(),
```
It's great to see that multiple-output pattern replacement is working!

ProExpertProg commented on vllm/compilation/pass_manager.py:
```diff
@@ -31,24 +33,29 @@ class PostGradPassManager(CustomGraphPass):
+        shape = get_pass_context().runtime_shape
         for pass_ in self.passes:
-            pass_(graph)
+            if pass_.is_applicable_for_shape(shape):
```
If this pass is an `InductorPass` instead of a `VllmInductorPass` (when added via `CompilationConfig.inductor_config` from outside vLLM), this method will not exist. We should either check this here or move the method to `InductorPass`

ProExpertProg commented on vllm/compilation/vllm_inductor_pass.py:
```diff
@@ -50,6 +54,9 @@ def end_and_log(self):
         duration_ms = float(self._end_time - self._start_time) / 1.0e6
         logger.debug("%s completed in %.1f ms", self.pass_name, duration_ms)
 
+    def is_applicable_for_shape(self, shape: Optional[int]):
```
Alternatively, we could move this method to `InductorPass` - I know it's a little confusing right now but `VllmInductorPass` is for internal custom passes and `InductorPass` is the general interface that can be used for passes. We should probably get rid of this distinction.

cascade812 replied:
moved it to `InductorPass`

ProExpertProg reviewed (commented):
A few notes on certain edge cases. We should at least add TODOs if we don't address them here

ProExpertProg commented on tests/compile/test_functionalization.py:
```diff
@@ -49,13 +49,14 @@ def test_fix_functionalization(model: str, quant_key: QuantKey,
 
     passes = [noop_pass, fusion_pass] if do_fusion else [noop_pass]
-    func_pass = FixFunctionalizationPass(config)
+    func_pass = FixFunctionalizationPass(vllm_config)
```
Should probably add a case with seq parallel if supported, to make sure those nodes get defunctionalized properly.

cascade812 replied:
sequence parallelism pass doesn't replace auto-functionalized nodes. I think it's not necessary to add it here.

ProExpertProg replied:
Doesn't it match and replace RMS nodes?

cascade812 replied:
updated

ProExpertProg commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -0,0 +1,100 @@
+    ]
+
+    outputs = llm.generate(prompts, sampling_params)
+    for output, answer in zip(outputs, answers):
```
I think this test would be better with the `TestBackend` in the same directory, which gives direct access to the `fx.Graph`. We could add the `PostGradPassManager` to the `TestBackend` if necessary. But it's better because it isolates the pass(es) and makes it easier to unit test. The end-to-end behavior should be covered with the test in the `tests/distributed` directory.

cascade812 replied:
@ProExpertProg I'm unable to get your response from slack. I tried to do the test same as 'test_functionalization.py'. However, it has below issue for communication ops. 
``` File "/vllm/vllm/distributed/parallel_state.py", line 802, in get_tp_group
    assert _TP is not None, ("tensor model parallel group is not initialized")
           ^^^^^^^^^^^^^^^
AssertionError: tensor model parallel group is not initialized```
If I put the initialization of `SequenceParallelismPass` after LLM initialization, it's able to get TP, but test is stuck there for unknown reason.
Also I tried to set up the test env same as `test_comm_ops.py`, it still doesn't work (it has some other problems).
Do you know how to solve above error properly?

cascade812 replied:
@ProExpertProg Updated this test based on your comments. Now it applies sequence parallelism pass on a test model and uses `TestBackend` to access graph before and after, following the same approach used in `test_fusion.py`

ProExpertProg commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -0,0 +1,100 @@
+        splitting_ops=[],
+    )
+    config.pass_config.enable_sequence_parallelism = True
+    config.pass_config.dump_graph_dir = Path(temp_dir)
+    config.pass_config.dump_graph_stages = \
```
We're actually planning on deprecating the `dump_graph_dir` in favor of depyf inspection, so tests really shouldn't depend on it.

ProExpertProg reviewed (commented):
Hey, sorry for the kinda-late review, I've been head-down on the attention-quant fusion pass.

This is a great PR! I had a few minor notes - I've submitted some above, and I had a few more notes on testing I thought of so submitting them now.

I'm happy to provide more explanation for the comments & discuss if I misunderstood something. Feel free to ping me on the vLLM Slack or here whether for discussion or if you've addressed the comments and want me to take another look.

cascade812 commented on vllm/config.py:
```diff
@@ -3826,14 +3829,17 @@ def __post_init__(self):
         if self.compilation_config is None:
             self.compilation_config = CompilationConfig()
+        if self.compilation_config.pass_config.enable_sequence_parallelism:
+            self.compilation_config.custom_ops.append("+rms_norm")
```
it can't be None. There's a validation error if it's None.

cascade812 committed:
address comments

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented on vllm/compilation/vllm_inductor_pass.py:
```diff
@@ -24,8 +25,11 @@ class VllmInductorPass(InductorPass):
-    def __init__(self, config: CompilationConfig.PassConfig):
-        self.config = config
+    def __init__(self, config: VllmConfig):
+        self.config = config.compilation_config.pass_config
```
updated

cascade812 commented on vllm/config.py:
```diff
@@ -3826,14 +3829,17 @@ def __post_init__(self):
             # and avoid any potential issues with the inductor.
             # FIXME(rob): Add function to set all of these.
-            self.compilation_config.custom_ops = ["none"]
+            if self.compilation_config.custom_ops is None:
```
updated

cascade812 commented on vllm/config.py:
```diff
@@ -3918,6 +3936,15 @@ def _set_cudagraph_sizes(self):
+                    self.compilation_config.pass_config.enable_sequence_parallelism:
+                    possible_sizes = [
+                        size for size in possible_sizes if size %
+                        self.parallel_config.tensor_parallel_size == 0
```
added

cascade812 committed:
minor fix

Signed-off-by: cascade812 <cascade812@outlook.com>

zou3519 commented on vllm/compilation/compiler_interface.py:
```diff
@@ -303,11 +305,12 @@ def _get_shape_env() -> AlwaysHitShapeEnv:
-                example_inputs,
-                inner_compile=hijacked_compile_fx_inner,
-                config_patches=current_config)
+            with pass_context(runtime_shape):
```
Do you have an explanation of how this works? What are do we end up doing with the pass context? (does something read from it?)

cascade812 replied:
it pass the runtime_shape as context which it reads later when decides if apply sequence parallelism pass or not.

zou3519 replied:
thanks, that makes sense to me

cascade812 committed:
update test

Signed-off-by: cascade812 <cascade812@outlook.com>

ProExpertProg reviewed (commented):
Had a few more comments and a bunch more nits, thanks for addressing all of the comments so far.

A big question is whether the `tp_world_size` can be different for two LLM instances, because this pass being a singleton means that it'll only use the tp_size that was present during the first initialization. @tlrmchlsmth do you know the answer to this?

I think we should be able to resolve this by checking the tp_world_size during the `instance()` call, and then we can register more patterns on the same instance or maybe even return a new instance with the new tp_size (if that works). Or, because we clear `_seen_patterns`, we could get rid of the singleton structure anyway (you'd have to verify it works).

cascade812 committed:
test  FixFunctionalizationPass  with SequenceParallelismPass

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented on tests/compile/test_sequence_parallelism.py:
```diff
@@ -0,0 +1,172 @@
+@pytest.mark.parametrize("seq_len", [16])
+@pytest.mark.parametrize("hidden_size", [16])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
+@pytest.mark.skipif(envs.VLLM_TARGET_DEVICE not in ["cuda"],
```
added fix_functionalization in the test

cascade812 committed:
remove redundant code

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented on vllm/compilation/sequence_parallelism.py:
```diff
@@ -0,0 +1,296 @@
+            return all_gather, reduce_scatter
+
+        def get_inputs():
+            arg2_1 = torch.rand([16, 4], device=self.device, dtype=self.dtype)
+            mul_6 = torch.tensor([[3, 7, 1, 4, 9, 2, 5, 0]],
+                                 device=self.device,
+                                 dtype=torch.long)
+            unsqueeze = torch.rand([1, 8, 1], device=self.device, \
+                dtype=self.dtype) > 0.5
+            full_default = torch.zeros([1, 8, 4], device=self.device, \
+                dtype=self.dtype)
+            permute = torch.rand([1, 8, 4],
+                                 device=self.device,
+                                 dtype=self.dtype)
+            arg3_1 = torch.rand([4], device=self.device, dtype=self.dtype)
+            return [arg2_1, mul_6, unsqueeze, full_default, permute, arg3_1]
```
updated some of them. The left two (`unsqueeze` and `mul_6 `) values do matter,  values of `mul_6` have to be less than the first dimension of arg2_1. And values of `unsqueeze` is to created a tensor of random bool values.

cascade812 committed:
Merge remote-tracking branch 'origin' into sp_pass

cascade812 committed:
remove the singleton pattern to support two LLM instances.

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented:
> Had a few more comments and a bunch more nits, thanks for addressing all of the comments so far.
> 
> A big question is whether the `tp_world_size` can be different for two LLM instances, because this pass being a singleton means that it'll only use the tp_size that was present during the first initialization. @tlrmchlsmth do you know the answer to this?
> 
> I think we should be able to resolve this by checking the tp_world_size during the `instance()` call, and then we can register more patterns on the same instance or maybe even return a new instance with the new tp_size (if that works). Or, because we clear `_seen_patterns`, we could get rid of the singleton structure anyway (you'd have to verify it works).

Good point,  I removed singleton and tested two llm instances. It works well.

cascade812 committed:
nit

Signed-off-by: cascade812 <cascade812@outlook.com>

cascade812 commented:
> Thanks for the great work! @cascade812 Just a simple question, I've noticed that you initialized the paramater rms_norm_weight with an empty 2D tensor during the middleRMSNorm and the lastRMSNorm SP pass, but on the CUDA kernel side, the function accepts it as a 1D tensor. Why wasn't it conflicted with each other?

I think it's because pattern matcher is more flexible, it doesn't validate kernel implementation details - it just ensures the operation pattern is recognizable and shapes are broadcastable.
